hwy/contrib/unroller/unroller_test.cc

```diff
@@ -0,0 +1,295 @@
+    auto maskneg = hn::Not(hn::FirstN(
+        d, static_cast<size_t>(places +
+                               static_cast<ptrdiff_t>(
+                                   UnrollerUnit<minv<T>, T, T>::UnitLanes()))));
+    if (places < 0) mask = maskneg;
+
+    auto def = YInitImpl();
+    return hn::MaskedLoadOr(def, mask, d, from + idx);
+  }
+
+  inline ptrdiff_t StoreImpl(ptrdiff_t const& HWY_RESTRICT idx,
+                             T* HWY_RESTRICT to,
+                             hn::Vec<TT> const& HWY_RESTRICT x) const {
+    // no stores in a reducer
+    return 0;
+  }
+
+  inline ptrdiff_t MaskStoreImpl(ptrdiff_t const& HWY_RESTRICT idx,
+                                 T* HWY_RESTRICT to,
+                                 hn::Vec<TT> const& HWY_RESTRICT x,
+                                 ptrdiff_t const& HWY_RESTRICT places) const {
+    // no stores in a reducer
+    return 0;
+  }
+
+  inline ptrdiff_t ReduceImpl(hn::Vec<TT> const& HWY_RESTRICT x,
+                              T* HWY_RESTRICT to) const {
+    const hn::ScalableTag<T> d;
+    auto minvect = hn::MinOfLanes(d, x);
+    (*to) = hn::ExtractLane(minvect, 0);
+    return 1;
+  }
+
+  inline void ReduceImpl(hn::Vec<TT> const& HWY_RESTRICT x0,
+                         hn::Vec<TT> const& HWY_RESTRICT x1,
+                         hn::Vec<TT> const& HWY_RESTRICT x2,
+                         hn::Vec<TT>& HWY_RESTRICT y) const {
+    auto a = hn::Min(x1, x0);
+    auto b = hn::Min(y, x2);
+    y = hn::Min(a, b);
+  }
+};
+
+template <typename T>
+struct dotv : UnrollerUnit2<dotv<T>, T, T, T> {
+  using TT = hn::ScalableTag<T>;
+
+  inline hn::Vec<TT> Func(ptrdiff_t const& HWY_RESTRICT idx,
+                          hn::Vec<TT> const& HWY_RESTRICT x0,
+                          hn::Vec<TT> const& HWY_RESTRICT x1,
+                          hn::Vec<TT> const& HWY_RESTRICT y) const {
+    return hn::MulAdd(x0, x1, y);
+  }
+
+  inline ptrdiff_t StoreImpl(ptrdiff_t const& HWY_RESTRICT idx,
+                             T* HWY_RESTRICT to,
+                             hn::Vec<TT> const& HWY_RESTRICT x) const {
+    // no stores in a reducer
+    return 0;
+  }
+
+  inline ptrdiff_t MaskStoreImpl(ptrdiff_t const& HWY_RESTRICT idx,
+                                 T* HWY_RESTRICT to,
+                                 hn::Vec<TT> const& HWY_RESTRICT x,
+                                 ptrdiff_t const& HWY_RESTRICT places) const {
+    // no stores in a reducer
+    return 0;
+  }
+
+  inline ptrdiff_t ReduceImpl(hn::Vec<TT> const& HWY_RESTRICT x, T* to) const {
+    const hn::ScalableTag<T> d;
+    (*to) = hn::ReduceSum(d, x);
+    return 1;
+  }
+
+  inline void ReduceImpl(hn::Vec<TT> const& HWY_RESTRICT x0,
+                         hn::Vec<TT> const& HWY_RESTRICT x1,
+                         hn::Vec<TT> const& HWY_RESTRICT x2,
+                         hn::Vec<TT>& HWY_RESTRICT y) const {
+    y = y + x0 + x1 + x2;
+  }
+};
+
+template <typename T>
```

Comment:
Let's move this to test_util.h?

Reply:
I don't quite understand how to do this yet.

Reply:
Ah, was just thinking: this pair of functions is now used in multiple tests. We can avoid the duplication by moving both of them to test_util.h, and adding HWY_INLINE before the void just to be sure.